src/checkpoint.hh

```diff
@@ -0,0 +1,432 @@
+             << endl;
+     stream << indent() << "          << \" Will retry at "
+                            "the next checkpoint...\\n\";" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl << endl;
+     dec_indent(); dec_indent();
+  }
+
+  void remove(Printer::Base &stream) {
+     inc_indent(); inc_indent();
+     stream << indent();
+     stream << "void remove() {" << endl;
+     inc_indent();
+     stream << indent() << "boost::filesystem::remove(table_path);" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl << endl;
+     dec_indent(); dec_indent();
+  }
+
+  void get_table_path(Printer::Base &stream) {
+    inc_indent(); inc_indent();
+    stream << indent() << "std::string get_table_path() {" << endl;
+    inc_indent();
+    stream << indent() << "return table_path.string();" << endl;
+    dec_indent();
+    stream << indent() << "}" << endl << endl;
+    dec_indent(); dec_indent();
+  }
+
+  void find_table(Printer::Base &stream) {
+     inc_indent(); inc_indent();
+     stream << indent() << "boost::filesystem::path find_table("
+            << "boost::filesystem::path path, const std::string &tname) {"
+            << endl;
+     inc_indent();
+     stream << indent() << "// find the table archive in the "
+            << "user-specified directory" << endl;
+     stream << indent() << "std::string curr_file;" << endl;
+     stream << indent() << "for (const auto &entry : "
+            << "boost::filesystem::directory_iterator(path)) {" << endl;
+     inc_indent();
+     stream << indent() << "curr_file = entry.path().string();" << endl;
+     stream << indent() << "if (curr_file.find(tname) != curr_file.npos) {"
+            << endl;
+     inc_indent();
+     stream << indent() << "return entry.path();" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl;
+     stream << indent() << "return boost::filesystem::path(\"\");" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl << endl;
+     dec_indent(); dec_indent();
+  }
+
+  void init(Printer::Base &stream) {
+     inc_indent(); inc_indent(); inc_indent();
+     stream << indent() << "std::string executable_name = "
+            << "boost::dll::program_location().filename().string();" << endl;
+     stream << indent() << "int process_id = getpid();" << endl;
+     stream << indent() << "std::string archive_name = executable_name + "
+            << "\"_\" + std::to_string(process_id);" << endl;
+     stream << indent() << "table_path = out_path / (archive_name + \"_\" + "
+            << "tname);" << endl;
+     stream << indent() << "std::string in_path_error_msg = "
+            << "\"at user-specified path\";" << endl << endl;
+     stream << indent() << "if (in_path.empty()) {" << endl;
+     inc_indent();
+     stream << indent() << "in_path = "
```

Comment:
we should let the user know (write an Info msg) that a run could take a short-cut by loading a checkpoint

Reply:
Done.